spec/controllers/samples_controller_spec.rb

```diff
@@ -0,0 +1,112 @@
+require 'rails_helper'
+
+RSpec.describe SamplesController, type: :controller do
+  create_users
+
+  pipeline_run_stages_data = [{
+    name: "Host Filtering",
+    dag_json: "{\"key1\": \"value1\"}"
+  }, {
+    name: "GSNAPL/RAPSEARCH alignment",
+    dag_json: "{\"key2\": \"value2\"}"
+  }, {
+    name: "Post Processing",
+    dag_json: "{\"key3\": \"value3\"}"
+  }, {
+    name: "Experimental",
+    dag_json: "{\"key4\": \"value4\"}"
+  }]
+
+  # Admin specific behavior
+  context "Admin user" do
+    before do
+      sign_in @admin
+      @admin.add_allowed_feature("pipeline_viz")
```

Comment:
This should be done through factories... something like:
```
create(:joe)
@admin = create(:admin, allowed_features: ["pipeline_viz"])
sign_in_admin
```
(instead of using the create_users macro)
Rule of thumb is we should not call inner methods to change state here but inside of factories.